---
title: "2020_04_13_Sandbox_Code"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd('~/igraph_Network_Expansion/')
library(igraph)
library(paxtoolsr)
library(RJSONIO)
library(httr)
library(DOSE)
library(org.Hs.eg.db)
library(clusterProfiler)
library(parallel)
library(doParallel)
library(RColorBrewer)
#devtools::install_github("jhrcook/HotNetvieweR")
library(HotNetvieweR)
source('igraphhack/igraphplot2.R')
```

## R Import Data

```{r Import}
setwd('~/igraph_Network_Expansion/')

User <- '< USER_ID >'
Pass <- '< Password >'

reticulate::use_python("/usr/bin/python", required = TRUE)
synapseclient <- reticulate::import("synapseclient")
syn_temp <- synapseclient$Synapse()
syn_temp$login( User, Pass, '--rememberMe' )

rm(User)
rm(Pass)

SIFs <- c('syn21914063',
  'syn21914056',
  'syn21914057',
  'syn21914059',
  'syn21914061',
  'syn21914064',
  'syn21914065',
  'syn21914070',
  'syn21914071',
  'syn21914072',
  'syn21914078',
  'syn21914082',
  'syn21914083',
  'syn21914087',
  'syn21914090',
  'syn21914091',
  'syn21914093',
  'syn21914094',
  'syn21914074'
)

names(SIFs) <- c( 'Detailed', 
                  'bind', 
                  'biogrid', 
                  'corum', 
                  'ctd',
                  'dip',
                  'drugbank',
                  'hprd',
                  'humancyc',
                  'intacsyngnc',
                  'msigdb',
                  'netpath',
                  'panther',
                  'pathbank',
                  'pid',
                  'psp',
                  'reactome',
                  'reconx',
                  'kegg'
  )

#Load and Add names: SIFs
boot <- function( i ){
  foo <- read.table( file=syn_temp$get( SIFs[i] )$path , header = F, sep='\t' )
  foo$Pathway <- as.character( i ) 
  return(foo)
}

Loader <- function( SIFs){
  #'@SIFs An object that has SynID values and named for the database
  ##  Each Syn ID is a SIF format file 
  
  total <- list()
  Trial <- for( i in names(SIFs) ){
    temp <- boot( i )
    #eval(parse(text=paste0( 'total$', as.character(NAMES[i]), ' <- temp' )))
    total[[i]] <- temp
  }
  
  Total <- do.call(rbind, total)
  Total <- Total[ , c("V1", "V3", "V2", "Pathway") ]
  colnames(Total) <- c("from", "to", "interaction", "pathway")
  
  #Caluclate how many times this intereactcion was found in all databases:
  Total$Occurance <- paste0( Total$from, '-', Total$to, ':', Total$interaction )
  foo <- paste0( Total$from, '-', Total$to, ':', Total$interaction )
  TabFoo <- table(foo)
  Total$Occurance <-  as.numeric( TabFoo[ Total$Occurance ] )
  
  Genes <- c( as.character(Total$from), as.character(Total$to) )
  Genes <- Genes[ !duplicated(Genes) ]
  Genes <- as.data.frame( Genes )
  
  #Make the pathway column into a list object
  #library(dplyr)
  #library(reshape)
  
  library(data.table)
  Total$UniqCol <- paste0( as.character(Total$from),':', as.character(Total$to),'-', as.character(Total$interaction) )
  dt <- data.table(Total[, c('UniqCol','pathway')])
  DT <- dt[,lapply(.SD, function(col) paste(col, collapse=", ")), by=.(UniqCol)]
  
  sinl<-DT
  
  foo <- as.data.frame( DT )
  poo <- as.list( strsplit(as.character(foo$pathway),',') )
  names(poo) <- foo$UniqCol
  
  totals <- Total[ !duplicated(Total$UniqCol), ]
  pathways <- poo
  
  table(names(pathways) == as.character(totals$UniqCol))
  table( as.character(totals$UniqCol) == names(pathways) )
  
  totals$PATH <- pathways
  totals$PATH <- lapply( totals$PATH,
                 function(x) gsub(" ","", x)
               )
  
  Total <- totals[,c("from", "to", "interaction", "Occurance", "UniqCol", "PATH")]
  colnames(Total) <- c("from", "to", "interaction", "Occurance", "UniqCol", "pathway")
  
  #Make into a network graph
  #Make into a network graph
  
  Graph <- list()
  for(type in levels(Total$interaction) ){
    #NET <- graph_from_data_frame(d=Total[ Total$interaction == type, ], vertices=Genes, directed=T) 
    #set_graph_attr(NET,"interaction","controls-transport-of-chemical")
    eval( parse( text=paste0('Graph$`', type, '` <- graph_from_data_frame(d=Total[ Total$interaction == \'', type, '\', ], vertices=Genes, directed=T) ') ) )
  }
  #Edge Sum is 1903389
  
  #Another way to make the Network object But has multiple edges per-Vertex set
  net_oldStyle <- graph_from_data_frame(d=Total, vertices=Genes, directed=T) 
  return( list( DF=Total, NET=net_oldStyle, Gra=Graph ) )
}

Data <- Loader(SIFs)
Total <- Data$DF
net_oldStyle <- Data$NET
Graph <- Data$Gra

Tot <- 0
for(type in levels(Total$interaction) ){
  message( paste0( type, ": ", eval(parse(text=paste0(  'length(E(Graph$`', type, '`))' ))) ))
  Tot <- eval(parse(text=paste0( 'length(E(Graph$`', type, '`))') ))+Tot
}
message( paste0( "Total Interactions = ", Tot ) )


#_# Test Code
#@sk:All interactions of our interested partners
Total[ (Total$from =='PIK3R1' & Total$to=='PIK3CA') | (Total$from =='PIK3CA' & Total$to=='PIK3R1') , ]

#@sk:Do we have multiple edge IDs:
table( which_multiple(net_oldStyle, eids = E(net_oldStyle)) )
#   FALSE    TRUE 
# 1804977   98412
#---# Yes We Do

#@sk:
fun<-Total
fun$Extra <- paste0( as.character(fun$from), ':', as.character(fun$to) )
table(table(fun$Extra))

#Get the shortest path...
remove <- names(V(net_oldStyle))[ grepl( 'CHEBI:', names(V(net_oldStyle)) ) ==T ]
Slim <- delete_vertices( net_oldStyle, remove )

news.path <- shortest_paths( Slim, 
                                from = V(Slim)["SMOC1"], 
                                to  = V(Slim)["SORL1"],
                                mode = "all",
                                output = c('both')
                            )# both path nodes and edges

Trial.path <- induced_subgraph( net_oldStyle, colnames(do.call( rbind, news.path$vpath )) )
igraph::plot.igraph(Trial.path, edge.arrow.size=.4)

# Get the vertecies of N Gene Ids
####################################################################################
MoesinScores <- data.table::fread(syn_temp$get('syn21652325')$path)
MoesinGenes <- MoesinScores$GeneName
GeneSetA <- c( "A1BG", "ABCC5", "ABI1", "ABI1", "ABI2", "ACACA", "PIK3R1", "PIK3CA" )
GeneSetB <- c("CD44", "FLT1", "NTN1", "SFRP1", "STX4", "LACTB", "PVR", "ACE", "EPHX2", "RABEP1", "MADD", "APP", "MAPT")

#Subgraph:
Trial.path <- induced_subgraph( net_oldStyle, GeneSetB )

GS <- GeneSetB
message( paste0( length( GS[ GS%in% names(V(net_oldStyle)) ] ), ' of ', length( GS ), ' genes found in the Network' ))

Trial.path <- induced_subgraph( net_oldStyle, GS[ GS%in% names(V(net_oldStyle)) ] )

#Trial.path <- E(net_oldStyle)[ V(net_oldStyle)[ names(V(net_oldStyle)) %in% GS] ]
Trial.path <- subgraph.edges( net_oldStyle, E(net_oldStyle)[ .inc(V(net_oldStyle)[GeneSetB])] )

#Remove all CHEBI: nonsense
Trial.path <- subgraph.edges( Trial.path, E(Trial.path)[ grepl( 'CHEBI', edge_attr(Trial.path)$UniqCol)  == F ] )


##_#
foo <- subgraph.edges( net_oldStyle, E(net_oldStyle)[ grepl( 'CHEBI', edge_attr(net_oldStyle)$UniqCol)  == F ] )

#############################################################################################################################
######Cluster the Total Graph
#devtools::install_github("igraph/rigraph")

#Potential Weights For Clustering....
Total$EdgeRep <- 0
Total$Edge <- paste0( Total$from, ':',  Total$to )
key <- table( Total$Edge )
Total$EdgeRep <- key[ Total$Edge ]

Total_Occurence <- rep( 0, length(names(key)) )
names( Total_Occurence ) <- names(key)

SumTally <- function( name,df ){
  Value <- sum( df[ df$Edge == name, ]$Occurance )
  return(c(name,Value))
}

SumTally <- function( name,df ){
  Value <- sum( df[ df$Edge == name, ]$Occurance )
  return(c(name,Value))
}

temp_Total <- Total[ ,c('Edge','Occurance')]
temp_Total$Occurance <- as.numeric(temp_Total$Occurance)

cl <- makeCluster( detectCores()-2 )
registerDoParallel(cl)

L <- length(names(Total_Occurence))
mark<-Sys.time()
#TOs <- foreach( i=names(Total_Occurence)[1:L], .combine=rbind, .export = ls(globalenv()) ) %dopar% SumTally( i,temp_Total )
TOs <- foreach( i=names(Total_Occurence)[1:L], .combine=rbind ) %dopar% SumTally( i,temp_Total )
Sys.time()-mark
# Time difference of 2.260386 days

stopCluster(cl)
#TOs <- read.csv('../ReferenceData/Occurance.csv', row.names=1)
sink_TO <-TOs

#write to file
#write.csv( temp_Total, file='../ReferenceData/Occurance.csv')
#write.csv( TOs, file='../ReferenceData/SumTalliedOccur.csv')
#TOs <- dplyr::distinct(TOs)
row.names(TOs) <- TOs[,1]
TOs <- as.data.frame(TOs)
TOs$V2 <- as.numeric(as.character(TOs$V2))
head(TOs[ Total$Edge, ])

Total$SumOccurence <- TOs[ Total$Edge, ]$V2

Genes <- c( as.character(Total$from), as.character(Total$to) )
Genes <- Genes[ !duplicated(Genes) ]
Genes <- as.data.frame( Genes )

Graph <- list()
for(type in levels(Total$interaction) ){
    eval( parse( text=paste0('Graph$`', type, '` <- graph_from_data_frame(d=Total[ Total$interaction == \'', type, '\', ], vertices=Genes, directed=T) ') ) )
}
#Edge Sum is 1903389
#save(Total, file = "../ReferenceData/Final_Total.tsv")
#load("../ReferenceData/Final_Total.tsv")
#Another way to make the Network object But has multiple edges per-Vertex set
net_oldStyle <- graph_from_data_frame(d=Total, vertices=Genes, directed=T) 
Data <- list( DF=Total, NET=net_oldStyle, Gra=Graph )

```

```{r PullCoExp}

#############################################################################################################################
#Add coexpression to the

library("biomaRt")
#dim(Total[ grepl('CHEBI', as.character(Total$UniqCol)) == F, ])
Total <- Total[ grepl('CHEBI', as.character(Total$UniqCol)) == F, ]

mart = useDataset("hsapiens_gene_ensembl", useEnsembl(biomart="ensembl", version=84))

Genes <- c( as.character(Total$from), as.character(Total$to) )
Genes <- Genes[ !duplicated(Genes) ]

TRANS <- getBM(attributes=c('hgnc_symbol', 'ensembl_gene_id'), 
      filters = 'hgnc_symbol', 
      values = Genes, 
      mart = mart)

library(dplyr)
Pull_CoExp <- function( Data, SynID, trans, DF){
  #'@Data Character vector of the Brain Region eg. 'DLPFC'
  #'@SynID SynID of the coexpression matrix eg. 'syn21630206'
  #'@trans gene name to ensg dataframe for DF eg. TRANS
  #'@DF The network dataframe eg Total
  
  #Data <- 'DLPFC'
  #SynID <- 'syn21630206'
  #trans <- TRANS
  #DF <- Total[ , colnames(Total) != 'DLPFC_CE' ]
  
  #DLPFC
  CoExp <- read.table(syn_temp$get(SynID)$path, header=T,sep='\t')
  
  #Filter trans for relevant ENSGs:
  trans <- trans[ trans$ensembl_gene_id %in% colnames(CoExp), ]
  
  #Filter CoExp for relevant Genes:
  CoExp_filt <- CoExp[ row.names(CoExp)[ row.names(CoExp) %in% trans$ensembl_gene_id],
                       colnames(CoExp)[ colnames(CoExp) %in% trans$ensembl_gene_id]
                       ]
  #Make the Statistic Column:
  eval(parse(text= paste0( 'DF$', Data, '_CE <- NA' ) ))
  
  #Split DF into unanotatable and need to be annotate:
  Total_UA <- DF[ (as.character(DF$from) %in% trans$hgnc_symbol) == F | (as.character(DF$to) %in% trans$hgnc_symbol) == F, ]
  Total_A <- DF[ (as.character(DF$from) %in% trans$hgnc_symbol) == T & (as.character(DF$to) %in% trans$hgnc_symbol) == T, ]
  
  # Melt the CoExpressed Filtered
  row.names(trans) <- trans$ensembl_gene_id
  row.names(CoExp_filt) <- trans[ row.names(CoExp_filt), ]$hgnc_symbol
  colnames(CoExp_filt) <- trans[ colnames(CoExp_filt), ]$hgnc_symbol
  
  m<-as.matrix(CoExp_filt)
  foo<-data.frame(col = rep(colnames(m), each = nrow(m)), 
             row = rep(rownames(m), ncol(m)), 
             value = as.vector(m))
  
  #Filter the coexpression for the pairs that matter ie not zero
  foo %>% mutate_if(is.factor, as.character) -> foo2
  
  Final_Annote <- foo[ foo$value > 0,]
  Final_Annote %>% mutate_if(is.factor, as.character) -> Final_Annote
  
  temp<-Final_Annote
  temp <- temp[ temp$col != temp$row, ]
  temp <- temp[,c(2,1,3)]
  colnames(temp) <- c( 'col', 'row', 'value')
  
  Exp <- merge( Final_Annote, temp )
  row.names(Exp) <- paste0( Exp$row, ':', Exp$col )
  
  #Split the Total_A (Annotate into zero and non-zero)
  Total_Z <- Total_A[ (Total_A$Edge %in% row.names(Exp)) == F, ]
  eval(parse(text=paste0( 'Total_Z$', Data,'_CE <- 0' ) ))
  
  Total_A <- Total_A[ Total_A$Edge %in% row.names(Exp), ]
  eval(parse(text=paste0( 'Total_A$', Data, '_CE <- Exp[ Total_A$Edge, ]$value') ))
  
  #Merge the datasets
  Total_CE <- dplyr::bind_rows( Total_A,Total_Z )
  DF <- dplyr::bind_rows( Total_CE, Total_UA )
  return(DF)
}

CoExpression <- c( 'syn21630206', 'syn21592277', 
                   'syn21655283', 'syn21965435', 
                   'syn21785669', 'syn21911919', 
                   'syn21898644' )
names( CoExpression ) <- c( 'DLPFC', 'CBE', 
                            'FP', 'IFG', 
                            'PHG', 'STG', 
                            'TCX' )

# Run accross all regions:
for( i in 1:length(CoExpression) ) {
  Total <- Pull_CoExp( names( CoExpression )[i], CoExpression[i], TRANS, Total )
}

sink_Total <- Total

Total$Avg_Cortex_CE <- apply( Total[,c('DLPFC_CE', 'FP_CE', 'IFG_CE', 'PHG_CE', 'STG_CE', 'TCX_CE')], 1, mean )

Total$Avg_All_CE <- apply( Total[,c('DLPFC_CE', 'CBE_CE', 'FP_CE', 'IFG_CE', 'PHG_CE', 'STG_CE', 'TCX_CE')], 1, mean )

sink_Total <- Total
#save( Total, file='Total_Net_wCoexpression.RData' )
#save.image('2020_08_13_CoexpAdd.RData')

```

```{r MakeNet }
#Try Cortex CE net:
#Implement the edge_betweeness Algo:
Total_Cor <- Total[ Total$Avg_Cortex_CE > 0,]
#Total_Cor <- Total[ !is.na(Total$Avg_Cortex_CE) ,]
Genes <- c( as.character(Total_Cor$from), as.character(Total_Cor$to) )
Genes <- Genes[ !duplicated(Genes) ]

# Cluster EDGE BETWEENESS CLUSTERING
##Directed Graph:

######TEST
#CACNA2D2:CACNG2
Total_Cor <- Total[ !is.na(Total$Avg_Cortex_CE),]
Total_Cor <- Total_Cor[ Total_Cor$Avg_Cortex_CE > 0,]
#Total_Cor <- Total_Cor[1:1000,]
Total_Cor$from <- as.factor(as.character(Total_Cor$from))
Total_Cor$to <- as.factor(as.character(Total_Cor$to))
Genes <- c( as.character(Total_Cor$from), as.character(Total_Cor$to) )
Genes <- Genes[ !duplicated(Genes) ]

Total_Cor$from <- as.character(Total_Cor$from)
Total_Cor$to <- as.character(Total_Cor$to)
head( apply( Total_Cor[,c('to','from')], 1, sort ) )
Total_Cor$altUnique <- paste0( as.character(Total_Cor$to), ':', as.character(Total_Cor$from), '-', as.character(Total_Cor$interaction) )
#Total_Cor <- Total_Cor[ (Total_Cor$altUnique %in% Total_Cor$UniqCol)==F ,]
#Total_Cor[ grepl( 'CACNA2D2:CACNG2', as.character(Total_Cor$UniqCol)) | grepl( 'CACNG2:CACNA2D2', as.character(Total_Cor$UniqCol)), ]

#Remove reciprocate duplicates:
tmp <- t(apply( Total_Cor[,c('to','from')], 1, sort ) ) 
Total_Cor$altUnique <- paste0( as.character(tmp[,1]), ':', as.character(tmp[,2]), '-', as.character(Total_Cor$interaction) )
Total_Cor <- Total_Cor[ !duplicated(Total_Cor$altUnique), ]
Total_Cor <- Total_Cor[ , colnames(Total_Cor) != 'altUnique' ]

#Make iGraph net and cluster
net_undirected <- graph_from_data_frame(d=Total_Cor, vertices=Genes, directed=F) 
# Louvain Clustering
Start <- Sys.time()
Clust_bt <- cluster_louvain(net_undirected, weights = E(net_undirected)$Avg_Cortex_CE)
Sys.time()-Start

#plot(net_undirected, mark.groups = Clust_bt)


#James-Stein Estimator
#install.packages("corpcor")
library(corpcor)
foo <- as_adjacency_matrix(net_undirected, type = "upper", attr = "Avg_Cortex_CE", edges = FALSE, names = TRUE)#, sparse = igraph_opt("sparsematrices"))

foo.shrink <- cor.shrink(foo)
#length(as.vector(foo)[as.vector(foo) >0])
# 22634
#length(as.vector(foo.shrink)[as.vector(foo.shrink) >0])
# 272149
#length(as.vector(foo.shrink)[as.vector(foo.shrink) != 0])
# 61419829
#length(as.vector(foo.shrink))
#82755409
library(tidyr)
DT <- data.frame(col = rep(colnames(foo.shrink), each = nrow(foo.shrink)), 
             row = rep(rownames(foo.shrink), ncol(foo.shrink)), 
             value = as.vector(foo.shrink))
DT %>% dplyr::mutate_if(is.factor, as.character) -> DT
row.names(DT) <- paste0( DT[,1], ':', DT[,2] )

Total_Cor$JS <- abs( DT[ Total_Cor$Edge, ]$value )

#Make iGraph net and cluster
JS_net_undirected <- graph_from_data_frame(d=Total_Cor, vertices=Genes, directed=F) 
# Louvain Clustering
Start <- Sys.time()
JS_Clust_bt <- cluster_louvain(JS_net_undirected, weights = E(JS_net_undirected)$JS)
Sys.time()-Start

compare( JS_Clust_bt, Clust_bt, method="vi")
compare( JS_Clust_bt, Clust_bt, method="nmi")
compare( JS_Clust_bt, Clust_bt, method="split.join")
compare( JS_Clust_bt, Clust_bt, method="rand")
compare( JS_Clust_bt, Clust_bt, method="adjusted.rand")

##Look at Jesse's Kinases:
Small <- c( 'AXL', 'CERK', 'NME3', 'AATK', 'SRC', 'LTK', 'JAK1', 'PKCA' )
small <- Small[ Small %in% names(V(JS_net_undirected)) ]
#names(V(JS_net_undirected))[ grepl( 'PKC', names(V(JS_net_undirected)) ) ]

Big <- c("AXL", "CERK", "TSSK3", "AATK", "NME3", "PANK4", "SNX16", "STRADB", "JAK2", "NEK11", "MAST1", "MAPKAPK3", "MINK1", "PKMYT1", "TESK2", "CSNK1G3", "EPHA10", "TIE1", "TRRAP", "AGK", "SIK3", "GSK3A", "MELK", "MAP4K5", "DYRK4", "PRKCA", "NEK1", "DYRK1A", "CSF1R", "GK2", "MAGI2", "MAST2", "TESK1", "MAP3K4", "BCKDK", "SRC", "AURKB", "NTRK3", "MST1R", "STK11", "FGR", "PLXNA1", "MAPKAPK5", "JAK1", "LTK")
big <- Big[ Big %in% names(V(JS_net_undirected)) ]
#Big[ (Big %in% names(V(JS_net_undirected)))==F ]

#Expand the big geneset:
temp <- DT[ (DT$col %in% big | DT$row %in% big) & DT$value > 0.001, ]
poo <- c(temp$col,temp$row)
poo <- poo[!duplicated(poo)]

#Non-James Stein
vect <- rep(NA, length(Clust_bt))
for( i in 1:length(Clust_bt) ){
  PVal <- pbinom( length( poo[ poo %in% Clust_bt[[i]] ] ), 
       size=length(Clust_bt[[i]] ), 
       prob=length(poo)/length( V(net_undirected)),
       lower.tail = F
  )
  vect[i] <- PVal
}
vect_cor <- p.adjust(vect, method='fdr', n=length(vect) )
message(paste0("Non-James-Stein Estimator:"))
message(paste0("Significant Modules: ", paste(which( vect_cor < 0.05 ), collapse = ', ')))
message(paste0("Significant Module Cor P-values: ", paste(vect_cor[vect_cor<0.05], collapse = ', ')))
#Annotate the network with module membership
V(net_undirected)$community <- Clust_bt$membership

size<-NULL
Exp_Genes <- NULL
OL_B <- NULL
OL_S <- NULL
for( i in 1:length(which( vect_cor < 0.05 )) ) {
  size <- c(size, length( Clust_bt[[ which( vect_cor < 0.05 )[i] ]] ))
  OL_B <- c(OL_B, length( Big[ Big %in% Clust_bt[[ which( vect_cor < 0.05 )[i] ]] ]) )
  if( length( Small[ Small %in% Clust_bt[[ which( vect_cor < 0.05 )[i] ]] ]) > 0 ){
    OL_S <- c(OL_S, length( Small[ Small %in% Clust_bt[[ which( vect_cor < 0.05 )[i] ]] ]))
  }else{
    OL_S <- c(OL_S,0)
  }
  Exp_Genes <- c(Exp_Genes, length( poo[ poo %in% Clust_bt[[ which( vect_cor < 0.05 )[i] ]] ]) )
}

#Reduce Network to Significantly enriched Clusters
Keep = as.vector(unlist( Clust_bt[ which( vect_cor < 0.05 ) ] ))
Trial.path <- induced_subgraph(net_undirected, Keep)

Enriched_Clustered_NoEstimator <- list( DataFrame     = Total_Cor,
      FullNet       = net_undirected,
      Clusters      = Clust_bt,
      CorectedPVals = vect_cor,
      Significant_ModuleNet = Trial.path )
save(Enriched_Clustered_NoEstimator, file = 'Enriched_Clustered_NoEstimator.RData')
igraph::write_graph(Trial.path, "~/igraph_Network_Expansion/Enriched_Clustered_NoEstimator.graphml", format = "graphml")
 
#####################################
###Expanded James-Stein Estimator set
Vect <- rep(NA, length(JS_Clust_bt))
for( i in 1:length(JS_Clust_bt) ){
  PVal <- pbinom(length( poo[ poo %in% JS_Clust_bt[[i]] ] ), 
       size=length(JS_Clust_bt[[i]] ), 
       prob=length(poo)/length( V(JS_net_undirected)),
       lower.tail = F
  )
  Vect[i] <- PVal
}

JS_Vect_cor <- p.adjust(Vect, method='fdr', n=length(Vect) )
message(paste0("James-Stein Estimator:"))
message(paste0("Significant Modules: ", paste(which( JS_Vect_cor < 0.05 ), collapse = ', ')))
message(paste0("Significant Module Cor P-values: ", paste(JS_Vect_cor[JS_Vect_cor<0.05], collapse = ', ')))

#Annotate the network with module membership
V(JS_net_undirected)$community <- JS_Clust_bt$membership

JS_size<-NULL
JS_Exp_Genes <- NULL
JS_OL_B <- NULL
JS_OL_S <- NULL
Mod_Keep <- NULL
for( i in 1:length(which( JS_Vect_cor < 0.05 )) ) {
  JS_size <- c(JS_size, length( JS_Clust_bt[[ which( JS_Vect_cor < 0.05 )[i] ]] ))
  if( length( JS_Clust_bt[[ which( JS_Vect_cor < 0.05 )[i] ]]) > 10  ){
    Mod_Keep <- c(Mod_Keep,  which( JS_Vect_cor < 0.05 )[i] )
    JS_OL_B <- c(JS_OL_B, length( Big[ Big %in% JS_Clust_bt[[ which( JS_Vect_cor < 0.05 )[i] ]] ]) )
    if( length( Small[ Small %in% JS_Clust_bt[[ which( JS_Vect_cor < 0.05 )[i] ]] ]) > 0 ){
      JS_OL_S <- c(JS_OL_S, length( Small[ Small %in% JS_Clust_bt[[ which( JS_Vect_cor < 0.05 )[i] ]] ]))
    }else{
      JS_OL_S <- c(JS_OL_S,0)
    }
    JS_Exp_Genes <- c(JS_Exp_Genes, length( poo[ poo %in% JS_Clust_bt[[ which( JS_Vect_cor < 0.05 )[i] ]] ]) )
  }
}

#Reduce Network to Significantly enriched Clusters
Keep = as.vector(unlist( JS_Clust_bt[ Mod_Keep ] ))
JS.Trial.path <- induced_subgraph(JS_net_undirected, Keep)

Enriched_Clustered_JS_Estimator <- list( DataFrame = Total_Cor,
      FullNet       = JS_net_undirected,
      Clusters      = JS_Clust_bt,
      CorectedPVals = JS_Vect_cor,
      Significant_ModuleNet = JS.Trial.path )
save(Enriched_Clustered_JS_Estimator, file = 'Enriched_Clustered_JS_Estimator.RData')

igraph::write_graph(JS.Trial.path, "~/igraph_Network_Expansion/Enriched_Clustered_JS_Estimator.graphml", format = "graphml")
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
parentId = 'syn20798011';
activityName = 'Covariate and Diagnosis Regression';
activityDescription = 'Covariate analysis and Regrsison of AD Diagnosis of aligned effective counts with GRCh37 with CQN normalisation (IFG, STG, FP, PHG)';
thisFileName <- '2020_04_13_SandboxCode.Rmd'
# Github link
#devtools::install_github('brian-bot/githubr')
thisRepo <- githubr::getRepo(repository = 'jgockley62/igraph_Network_Expansion', ref="branch", refName='test')

testRepo <- githubr::getRepo(repository = 'brian-bot/githubr', ref="branch", refName='master')

repoCommit <- githubr::getRepo('jgockley62/igraph_Network_Expansion', ref='commit', refName='43b5d4d582e3343d0858d25ed041d83142d30812')
repoHead <- githubr::getRepo('jgockley62/igraph_Network_Expansion')

thisFile <- githubr::getPermlink(repository = thisRepo, repositoryPath=paste0('code/',thisFileName))
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
parentId <- 'syn21534582'
activityName = 'Igraph_KinaseNets';
activityDescription = 'Pilot of the igraph pathway tracing networks with Jesse\'s Kinase List';
CODE <- syn_temp$store(synapseclient$Folder(name = "igraphData", parentId = parentId))
#Set Used SynIDs For Provenance
Syns_Used <- c(  'syn21914063', 'syn21914056', 'syn21914057', 'syn21914059', 'syn21914061',
                 'syn21914064', 'syn21914065', 'syn21914070', 'syn21914071', 'syn21914072',
                 'syn21914078', 'syn21914082', 'syn21914083', 'syn21914087', 'syn21914090',
                 'syn21914091', 'syn21914093', 'syn21914094', 'syn21914074', 'syn21630206',
                 'syn21592277', 'syn21655283', 'syn21965435', 'syn21785669', 'syn21911919', 
                 'syn21898644' )
# Set annotations
all.annotations = list(
  dataType = 'Network',
  summaryLevel = 'gene',
  assay	 = 'RNAseq',
  tissueTypeAbrv	= c('IFG', 'STG', 'FP', 'PHG', 'TCX', 'DLFPC'), 
  study = c( 'MSBB', 'ROSMAP', 'Mayo' ), 
  organism = 'HomoSapiens',
  consortium	= 'TreatAD',
  genomeAssemblyID = 'GRCh38'
)
# Store JS Estimator - executed = thisFile, 
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='~/igraph_Network_Expansion/Enriched_Clustered_JS_Estimator.graphml', name = 'Clustered_JS_Estimator-Kinase List', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, activityDescription = activityDescription)
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
# Store Non JS Estimator - executed = executed = thisFile,
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='~/igraph_Network_Expansion/Enriched_Clustered_NoEstimator.graphml', name = 'Clustered_No_Estimator-Kinase List', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, activityDescription = activityDescription)
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
 
```

```{r plotcode}
#################### 
## Get subgraph & plot
Keep = names(V(JS_net_undirected))[ names(V(JS_net_undirected)) %in% JS_Clust_bt[[131]] ]#c( JS_Clust_bt[[N]], JS_Clust_bt[[131]]) ]
Trial.path <- induced_subgraph(JS_net_undirected, Keep)

#JS
Plot.NetWork <- JS.Trial.path
Basic_Net <- JS_net_undirected
#Non JS
Plot.NetWork <- Trial.path
Basic_Net <- net_undirected

#igraph::plot.igraph(JS_Clust_bt, as.undirected(Trial.path))
igraph::plot.igraph(Trial.path, edge.arrow.size=.4)

#Color the Edges Based on interaction type
Inters <- E(JS_net_undirected)$interaction
Inters <- Inters[ !duplicated(Inters) ]

n <- length(Inters)
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
pie(rep(1,n), col=sample(col_vector, n))

Colors <- sample(col_vector, n)
names(Colors) <- Inters

# Use Verticies
MiniNet <- Trial.path #induced.subgraph( Trial.path, as_ids(news.path$vpath[[1]]) )
MiniNet <- MiniNet %>%
  set_edge_attr("color", value = c( Colors[ c(E(Trial.path)$interaction) ]) )

plot( MiniNet, edge.arrow.size=.45 )

####Color the Edges Based on INTERACTION TYPE, EDGE OCCURENCE AMONG THE USER SELECTED PATHWAYS,

  #Choose Colors
Inters <- E(JS_net_undirected)$interaction
Inters <- Inters[ !duplicated(Inters) ]

n <- length(Inters)
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
pie(rep(1,n), col=sample(col_vector, n))

Colors <- sample(col_vector, n)
names(Colors) <- Inters

  ## - User input option to subgraph based on Pathway(s)
toMatch <- c( "Detailed", "bind", "biogrid", "corum", "ctd", "dip", "drugbank", "hprd", "humancyc", "intacsyngnc", "msigdb", "netpath", "panther", "pathbank", "pid", "psp", "reactome", "reconx", "kegg" )

Keeps <- grep( paste(toMatch,collapse="|"), E(Trial.path)$pathway)
Trial.path_Filt <- subgraph.edges(Trial.path, eids=Keeps, delete.vertices = TRUE)

Weights <- apply( sapply( E( Trial.path_Filt )$pathway, function(x) toMatch %in% x),
       2,
       function(x) table(x)["TRUE"]
)

  ## - Set Color and Weight Atributes
#MiniNet <- Trial.path_Filt #induced.subgraph( Trial.path, as_ids(news.path$vpath[[1]]) )
MiniNet <- Trial.path
MiniNet <- MiniNet %>%
  # Old for pre-pathway selection - set_edge_attr("weight",  value = lengths(E(Trial.path_Filt)$pathway )/(2*max( lengths( E(Trial.path_Filt)$pathway ))) ) %>%
  set_edge_attr("weight",  value = Weights /(1.25**max( lengths( Weights ) )) ) %>% 
  set_edge_attr( "color", value = c( Colors[ c(E(Trial.path_Filt)$interaction) ]) ) 
  
  ## Remove Direction from non-directional nodes
  E(MiniNet)$interaction
  Ints <- c( 0.4, 0.4, 0.4, 0.4,  0.4, 0.4, 0.4, 0.4, 0.4, 0.000001, 0.000001, 0.000001, 0.000001  )
  names(Ints) <- names( table(Total$interaction) )  
  DirectionArrowSize <- as.numeric( Ints[ E(MiniNet)$interaction ] )
  
  MiniNet <- MiniNet %>%
    set_edge_attr("DirectionArrowSize",  value = DirectionArrowSize ) 
  
  ## - Plot Final Network
par(xpd=TRUE)
environment(plot.igraph2) <- asNamespace('igraph')
environment(igraph.Arrows2) <- asNamespace('igraph')
plot.igraph2(MiniNet,edge.arrow.size=E(MiniNet)$DirectionArrowSize/max(E(MiniNet)$DirectionArrowSize)/2,
             edge.arrow.width=E(MiniNet)$DirectionArrowSize/max(E(MiniNet)$DirectionArrowSize),
             edge.width = E(MiniNet)$weight,
             vertex.size=5)
#plot( MiniNet, edge.width = E(MiniNet)$weight, edge.arrow.size= .4) #E(MiniNet)$DirectionArrowSize )   #.4 )E(graph)$direction)
legend('topleft', legend=names(Colors), 
       title.adj = 0.1,
       col= "black", pt.cex = 2, pch=21, 
       pt.bg=Colors, bty='n', title="Interaction Type" )
legend("topright", 
       #Old For All Pathways - legend = lengths(E(Trial.path_Filt)$pathway )[!duplicated(lengths(E(Trial.path_Filt)$pathway ))], #Labels
       legend = names(table( Weights )), #Labels
       lwd = as.numeric(names(table( Weights )))/(1.25*max( lengths( Weights ) )),
       col = "black",
       bty='n', 
       title="Pathway Redundancy")
mtext( paste0( "Pathways Examined: ", paste(toMatch, collapse = ', ')), 1 )


#################################################################
#Convert Network into Cytoscape Object:
source('~/igraph_Network_Expansion/NetworkAnalysis/cytoscape_util.R')

port.number = 8080
base.url = paste("http://localhost:", toString(port.number), "/v1", sep="")
base.url = paste("http://localhost:", toString(port.number),  sep="")

#base.url = paste("http://10.5.24.198:", toString(port.number), "/v1", sep="")

port.number = 8080
#base.url = paste("http://10.5.24.198:", toString(port.number), sep="")
base.url = paste("http://localhost:", toString(port.number),  sep="")
print(base.url)


### - NEED TO START CYTOSCAPE DOCKER CONTAINER AND NETWORK WITH Docker Compose  - ###
###Convert it into Cytosccape.js JSON
cygraph <- toCytoscape( MiniNet )
#[1] "Done.  To json Start..."
send2cy(cygraph, 'default%20black', 'circular')
#Error in curl::curl_fetch_memory(url, handle = handle) : 
#  Recv failure: Connection reset by peer


# https://github.com/cytoscape/cytoscape/releases/download/3.7.2/Cytoscape_3_7_2_unix.sh
# chmod u+x  Cytoscape_3_7_2_unix.sh

#No suitable Java Virtual Machine could be found on your system.
#The version of the JVM must be at least 1.8.0_152 and at most 9.
#Please define INSTALL4J_JAVA_HOME to point to a suitable JVM.

## java -version
# sh Cytoscape_3_7_2_unix.sh      # local installation
# sudo sh Cytoscape_3_7_2_unix.sh

############################# XXXXX #############################
#TRASH

#Make a network object (! This only has single edge types per-vertex !)
net <- igraph::disjoint_union( Graph$`interacts-with`,
               Graph$`in-complex-with`,
               Graph$`catalysis-precedes`,
               Graph$`chemical-affects`,
               Graph$`consumption-controlled-by`,
               Graph$`controls-expression-of`,
               Graph$`controls-production-of`,
               Graph$`controls-state-change-of`,
               Graph$`reacts-with`,
               Graph$`used-to-produce`,
               Graph$`controls-phosphorylation-of`,
               Graph$`controls-transport-of`,
               Graph$`controls-transport-of-chemical`
             )
```


```{r old code, eval=F}




Start <- Sys.time()
Clust_bt <- cluster_label_prop(net_directed, weights = E(net_directed)$Avg_Cortex_CE)
Sys.time()-Start


#Leading Eigen
Start <- Sys.time()
Clust_bt <- cluster_leading_eigen(simpnet_undirected, weights = E(simpnet_undirected)$Avg_Cortex_CE)
Sys.time()-Start
#Fast and greedy
net_undirected <- graph_from_data_frame(d=Total_Cor, vertices=Genes, directed=F) 

simpnet_undirected <- igraph::simplify(net_undirected, edge.attr.comb=list(interaction="ignore",
                                                                           Occurance="median",
                                                                           UniqCol="ignore",
                                                                           pathway="ignore",
                                                                           EdgeRep="median",
                                                                           Edge="first",
                                                                           SumOccurence="median"))

#Cluster Walk Trap
#Clust_bt <- cluster_walktrap(net_directed, weights = E(net_directed)$Avg_Cortex_CE, steps = 30, merges = TRUE, modularity = TRUE, membership = TRUE)
#27,729:  Time difference of 3.852172 secs
modularity(Clust_bt)
#[1] 0.6376457 - 4 steps
#[1] 0.7302233 - 15 steps
#[1] 0.7265575 - 30 steps
length(Clust_bt)
#[1] 1881 - 4 steps
#[1] 744  - 15 steps
#[1] 487 - 30 steps
length(ModSiz[ModSiz>10])
#[1] 84 - 4 steps
#[1] 59 - 15 steps
#[1] 30 - 30 steps

#louvain
#Clust_bt <- cluster_louvain(net_directed, weights = E(net_directed)$Avg_Cortex_CE)
#Edges
#27,729:      Time difference of 0.1426175 secs
modularity(Clust_bt)
#[1] 0.7683103
length(Clust_bt)
#[1] 174




#Betweeness
#Clust_bt <- cluster_edge_betweenness(net_directed, weights = E(net_directed)$Avg_Cortex_CE, directed = TRUE, edge.betweenness = TRUE, merges = TRUE, bridges = TRUE, modularity = TRUE, membership = TRUE)

#Edges
#100:      Time difference of 0.1548684 secs
#1,000:    Time difference of 1.530578 secs
#10,000:   Time difference of 5.353881 mins
#15,000:   Time difference of 17.65911 mins
#20,000:   Time difference of 46.51765 mins
#27,729:   Time difference of 2.439785 hours

ModSiz <- NULL
for( i in 1:length(JS_Clust_bt) ){
  ModSiz <- c(ModSiz, length(JS_Clust_bt[[i]]))
  #message( paste0( "Cluster ", i, " contains ", length(Clust_bt[[i]]), " Genes"))
}
modularity(Clust_bt)
length(ModSiz)
length( ModSiz[ ModSiz>10] )


########


#Total_Cor <- Total[ !is.na(Total$Avg_Cortex_CE) ,]
Total_Cor <- Total[ Total$Avg_Cortex_CE > 0,]
Genes <- c( as.character(Total_Cor$from), as.character(Total_Cor$to) )
Genes <- Genes[ !duplicated(Genes) ]

net_directed <- graph_from_data_frame(d=Total_Cor, vertices=Genes, directed=T) 

Clust_bt <- cluster_edge_betweenness(net_directed, weights = E(net_directed)$Avg_Cortex_CE, directed = TRUE, edge.betweenness = TRUE, merges = TRUE, bridges = TRUE, modularity = TRUE, membership = TRUE)


#net_undirected <- graph_from_data_frame(d=Total_Cor, vertices=Genes, directed=F) 

# Some Dependency not incl-Clust <- cluster_optimal(net_oldStyle, weights = NULL)
net_undirected <- graph_from_data_frame(d=Total[ Total$Avg_Cortex_CE > 0,], vertices=Genes, directed=F) 
# as.undirected(net_oldStyle)
simpnet_undirected <- igraph::simplify(net_undirected, edge.attr.comb=list(interaction="ignore",
                                                                           Occurance="median",
                                                                           UniqCol="ignore",
                                                                           pathway="ignore",
                                                                           EdgeRep="median",
                                                                           Edge="first",
                                                                           SumOccurence="median"))

# Remove vertices that have Chemical ID interaction
remove <- names(V(net_undirected))[ grepl( 'CHEBI:', names(V(net_undirected)) ) ==T ]
slim <- delete_vertices( simpnet_undirected, remove )

save(slim, file='../ReferenceData/Undirected_Network.RData')

# Fast/Greedy CLUSTERING
Clust_fg_EdgeRep <- cluster_fast_greedy(slim, merges = TRUE, modularity = TRUE, membership = TRUE, weights = E(slim)$Avg_Cortex_CE  )

Clust_fg_EdgeRep <- cluster_fast_greedy(slim, merges = TRUE, modularity = TRUE, membership = TRUE, weights = E(slim)$EdgeRep)

Clust_fg_SumOccurence <- cluster_fast_greedy(slim, merges = TRUE, modularity = TRUE, membership = TRUE, weights = E(slim)$SumOccurence)

Clust_fg_Occurance <- cluster_fast_greedy(slim, merges = TRUE, modularity = TRUE, membership = TRUE, weights = E(slim)$Occurance)

message( paste0( "Found ", length(Clust_fg_EdgeRep), " Clusters" ) )
message( paste0( "Found ", length(Clust_fg_SumOccurence), " Clusters" ) )
message( paste0( "Found ", length(Clust_fg_Occurance), " Clusters" ) )

for( i in 1:length(Clust_fg_Occurance) ){
  message( paste0( "Cluster ", i, " contains ", length(Clust_fg_Occurance[[i]]), " Genes"))
}

modularity( Clust_fg_Occurance )
# 0.2767961

Commun_fg <- igraph::communities( Clust_fg )

# EIGEN CLUSTERING
Clust_le_EdgeRep <- cluster_leading_eigen(slim, steps = -1, weights = E(slim)$EdgeRep, start = NULL, options = arpack_defaults, callback = NULL, extra = NULL, env = parent.frame())

Clust_le_SumOccurence <- cluster_leading_eigen(slim, steps = -1, weights = E(slim)$SumOccurence, start = NULL, options = arpack_defaults, callback = NULL, extra = NULL, env = parent.frame())

Clust_le_Occurance <- cluster_leading_eigen(slim, steps = -1, weights = E(slim)$Occurance, start = NULL, options = arpack_defaults, callback = NULL, extra = NULL, env = parent.frame())

#_#-LEFT ON FRIDAY HERE JKG

message( paste0( "Found ", length(Clust_le), " Clusters" ) )

for( i in 1:length(Clust_le) ){
  #message( paste0( "Cluster ", i, " contains ", length(Clust_le[[i]]), " Genes"))
}

modularity( Clust_le )
#0.2264285


# Cluster Walktrap CLUSTERING
Clust_wt <- cluster_walktrap(slim, weights = NULL, steps = 4,
  merges = TRUE, modularity = TRUE, membership = TRUE)

message( paste0( "Found ", length(Clust_wt), " Clusters" ) )

for( i in 1:length(Clust_wt) ){
 # message( paste0( "Cluster ", i, " contains ", length(Clust_wt[[i]]), " Genes"))
}

modularity( Clust_wt )
#0.2531823

# Cluster EDGE BETWEENESS CLUSTERING
Clust_bt <- cluster_edge_betweenness(slim, weights = NULL, directed = TRUE, edge.betweenness = TRUE, merges = TRUE, bridges = TRUE, modularity = TRUE, membership = TRUE)



#Clust_sg <- cluster_spinglass(simpnet_undirected, weights = NULL, vertex = NULL, spins = 25,
#  parupdate = FALSE, start.temp = 1, stop.temp = 0.01,
#  cool.fact = 0.99, update.rule = "random",
#  gamma = 1, implementation = "orig", gamma.minus = 1)



#-TOO INTENSIVE-girvan_newman <- cluster_edge_betweenness(net_oldStyle)
modularity(Clust)

Trial.path <- make_ego_graph(net_oldStyle,
                  order = 5,
                  nodes = V(net_oldStyle)[names(V(net_oldStyle))[ names(V(net_oldStyle)) %in% names(V(Trial.path)) ]],
                  mode = c("all"),
                  mindist = 0)

paste0( 'Trial.path[[', c(1:length(Trial.path)) , ']],' )


NewTrial <- paste0( paste0( 'Trial.path[[', c(1:(1-length(Trial.path))), ']] %u% '  ), Trial.path[[ length(Trial.path) ]]  )
NewTrial <- eval(parse( text=paste0( 'igraph::union(', 
                                     paste0( 'Trial.path[[', c(1:length(Trial.path)), ']],' ), 
                                     'byname = \'auto\')' )))
                        
                                     #paste0( 'Trial.path[[', (length(Trial.path)) , ']] )' )) ))


message( paste0( length(E(Trial.path)), ' direct edges found between ', length(V(Trial.path)), ' nodes'))

#Graph Object
Trial.OBJ <- make_ego_graph(net_oldStyle, order = 8, nodes = c( "APOE", "APOC1", "TREM2", "EED", "KNOP1", "CD2AP", "MTCH2", "CEACAM19", "CLPTM1" ), mode = c("all"), mindist = 0)

Grabber <- function( Input ){
  temp<-as_data_frame(Input, what = c("edges", "vertices", "both") )
  return( temp )                   
}

```